[8.x] Add ability to define default Password Rule

crynobone opened the pull request:
This will make it easier to define a generic default password rules for login, register and password reset and useful for Laravel UI, Breeze, Jetstream and Nova.

### Defining Rules

Typically on the `App\Providers\AuthServiceProvider`
```php
use Illuminate\Validation\Rules\Password;

Password::defaultUsing(function() {
     return Password::min(8)->uncompromised();
});
```

### Using the rules

E.g in Laravel Nova.

```php
use Laravel\Nova\Fields\Password;
use Illuminate\Validation\Rules\Password as PasswordRule;

Password::make('Password', 'password')->rules(PasswordRule::default());
```

Also, you can use:

```php
Password::required(); // return ['required', Password::min(8)->uncompromised()];

Password::sometimes(); // return ['sometimes', Password::min(8)->uncompromised()];
```

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
[8.x] Add ability to define default Password Rule.

This will make it easier to define a generic default password rules for login, register and password reset and useful for Laravel UI, Breeze, Jetstream and Nova.

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

driesvints commented on src/Illuminate/Validation/Rules/Password.php:
```diff
@@ -85,6 +93,31 @@ public function __construct($min)
+     *
+     * @return array
+     */
+    public static function fromDefault()
```
Maybe just:

```suggestion
    public static function default()
```

crynobone committed:
Update src/Illuminate/Validation/Rules/Password.php

Co-authored-by: Dries Vints <dries.vints@gmail.com>

crynobone committed:
update tests

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

driesvints commented:
Would be cool if you could do:

```php
Password::default()->required();
```

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented:
```php
Password::required()
```

I really liking this, `Password::required()` for login and reset password, `Password::sometimes()` for change password.

crynobone committed:
Rename assert method.

```
 1) Warning
Test method "testRule" in test class "Illuminate\Tests\Validation\ValidationPasswordRuleTest" is not public.
```

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
Add `Password::required()` and `Password::sometimes()`

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

taylorotwell commented:
So, if you use `Password::required`, I guess you are supposed to do something like this when building your validation rules?

```php
'name' => ['required', 'string'],
'password' => [...Password::required(), 'foo', 'bar'],
```

crynobone commented:
@taylorotwell yes, that's the plan with returning it as `array`.

Was tempting to do, but not sure if it's wise

```php
public static function __callStatic($method, $parameters)
{
    return [$method, static::default()];
}
```

lloricode commented:
Hi, did you just forget `confirmed` rule?

crynobone commented:
> Hi, did you just forget confirmed rule?

Intentionally leave it out, you can do `[...Password::required(), 'confirmed']` for specific use case.
